File: packages/cli/src/commands/sources.ts
import { Command } from 'commander'
import { Actionbook } from '@actionbookdev/sdk'
import chalk from 'chalk'
import { getApiKey, handleError, outputResult } from '../output.js'
import type { SourceItem } from '@actionbookdev/sdk'

export const sourcesCommand = new Command('sources')
  .description('List or search available sources (websites)')
  .option('-l, --limit <number>', 'Maximum results', '50')
  .option('-j, --json', 'Output raw JSON')
  .action(async (options) => {
    try {
      const apiKey = getApiKey(options)
      const client = new Actionbook({ apiKey })

      const result = await client.listSources({
        limit: parseInt(options.limit, 10),
      })

      if (options.json) {
        outputResult(result)
      } else {
        formatSourceList(result.results, result.count)
      }
    } catch (error) {
      handleError(error)
    }
  })

// Add search subcommand
sourcesCommand
  .command('search <query>')
  .alias('s')
  .description('Search for sources by keyword')
  .option('-l, --limit <number>', 'Maximum results', '10')
  .option('-j, --json', 'Output raw JSON')
  .action(async (query: string, options) => {
    try {
      // Get parent command options
      const parentOpts = sourcesCommand.opts()
      const apiKey = getApiKey({ ...parentOpts, ...options })
      const client = new Actionbook({ apiKey })

      const result = await client.searchSources({
        query,
        limit: parseInt(options.limit, 10),
      })

      if (options.json) {
        outputResult(result)
      } else {
        if (result.results.length === 0) {
          console.log(chalk.yellow(`\nNo sources found for "${query}"`))
        } else {
          console.log(chalk.bold.cyan(`\nSources matching "${query}"\n`))
          formatSourceList(result.results, result.count)
        }
      }
    } catch (error) {
      handleError(error)
    }
  })

function formatSourceList(sources: SourceItem[], count: number): void {
  if (sources.length === 0) {
    console.log(chalk.yellow('No sources found'))
    return
  }

  console.log(chalk.dim(`${count} source(s)\n`))

  // Calculate column widths
  const maxIdLen = Math.max(...sources.map((s) => String(s.id).length), 2)
  const maxNameLen = Math.min(Math.max(...sources.map((s) => s.name.length), 4), 30)

  // Header
  console.log(
    chalk.bold(
      `${'ID'.padEnd(maxIdLen)}  ${'Name'.padEnd(maxNameLen)}  ${'URL'}`
    )
  )
  console.log(chalk.dim('─'.repeat(80)))

  // Rows
  for (const source of sources) {
    const id = String(source.id).padEnd(maxIdLen)
    const name = truncate(source.name, maxNameLen).padEnd(maxNameLen)
    const url = source.baseUrl

    console.log(`${chalk.yellow(id)}  ${chalk.white(name)}  ${chalk.cyan(url)}`)

    if (source.description) {
      console.log(chalk.dim(`${''.padEnd(maxIdLen)}  ${truncate(source.description, 70)}`))
    }

    if (source.tags?.length) {
      console.log(
        chalk.dim(`${''.padEnd(maxIdLen)}  Tags: `) +
          source.tags.map((t) => chalk.magenta(t)).join(', ')
      )
    }
  }

  console.log()
  console.log(
    chalk.cyan('Tip: ') +
      chalk.white('Use --domain with search to filter by domain')
  )
  console.log(
    chalk.dim('Example: actionbook search "login" --domain github.com')
  )
}

function truncate(str: string, maxLen: number): string {
  if (str.length <= maxLen) return str
  return str.substring(0, maxLen - 1) + '…'
}

File: packages/actionbook-rs/src/commands/sources.rs
use colored::Colorize;

use crate::api::ApiClient;
use crate::cli::{Cli, SourcesCommands};
use crate::config::Config;
use crate::error::Result;

pub async fn run(cli: &Cli, command: &SourcesCommands) -> Result<()> {
    match command {
        SourcesCommands::List => list(cli).await,
        SourcesCommands::Search { query } => search(cli, query).await,
    }
}

async fn list(cli: &Cli) -> Result<()> {
    let mut config = Config::load()?;
    if let Some(ref key) = cli.api_key {
        config.api.api_key = Some(key.clone());
    }
    let client = ApiClient::from_config(&config)?;

    let response = client.list_sources(Some(50)).await?;

    if cli.json {
        println!("{}", serde_json::to_string_pretty(&response.results)?);
    } else {
        if response.results.is_empty() {
            println!("{} No sources available", "!".yellow());
            return Ok(());
        }

        println!("{} {} sources available\n", "✓".green(), response.count);

        for source in &response.results {
            println!(
                "{} {} {}",
                "●".cyan(),
                source.name.bold(),
                format!("(ID: {})", source.id).dimmed()
            );
            println!("  {} {}", "URL:".dimmed(), source.base_url);

            if let Some(ref desc) = source.description {
                if !desc.is_empty() {
                    let desc_preview = if desc.len() > 80 {
                        format!("{}...", &desc[..80])
                    } else {
                        desc.clone()
                    };
                    println!("  {} {}", "Desc:".dimmed(), desc_preview);
                }
            }

            if !source.tags.is_empty() {
                println!("  {} {}", "Tags:".dimmed(), source.tags.join(", "));
            }

            if let Some(score) = source.health_score {
                let score_color = if score >= 0.8 {
                    "●".green()
                } else if score >= 0.5 {
                    "●".yellow()
                } else {
                    "●".red()
                };
                println!("  {} Health: {:.0}%", score_color, score * 100.0);
            }

            println!();
        }
    }

    Ok(())
}

async fn search(cli: &Cli, query: &str) -> Result<()> {
    let mut config = Config::load()?;
    if let Some(ref key) = cli.api_key {
        config.api.api_key = Some(key.clone());
    }
    let client = ApiClient::from_config(&config)?;

    let response = client.search_sources(query, Some(20)).await?;

    if cli.json {
        println!("{}", serde_json::to_string_pretty(&response.results)?);
    } else {
        if response.results.is_empty() {
            println!(
                "{} No sources found matching: {}",
                "!".yellow(),
                query.bold()
            );
            return Ok(());
        }

        println!(
            "{} Found {} sources for: {}\n",
            "✓".green(),
            response.count,
            query.bold()
        );

        for source in &response.results {
            println!(
                "{} {} {}",
                "●".cyan(),
                source.name.bold(),
                format!("(ID: {})", source.id).dimmed()
            );
            println!("  {} {}", "URL:".dimmed(), source.base_url);

            if let Some(ref desc) = source.description {
                if !desc.is_empty() {
                    let desc_preview = if desc.len() > 80 {
                        format!("{}...", &desc[..80])
                    } else {
                        desc.clone()
                    };
                    println!("  {} {}", "Desc:".dimmed(), desc_preview);
                }
            }

            if !source.tags.is_empty() {
                println!("  {} {}", "Tags:".dimmed(), source.tags.join(", "));
            }

            println!();
        }
    }

    Ok(())
}
